# Question 1
In this question we're gonna find some association rules using apriori itemset selection algorithm. I've implemented the apriori and the rule extraction algorithms in the `apriori.py` file and imported the class here to use in the report.
We first define the transactions and initialize a new apriori solver instance.

In [1]:
from apriori import APriori

## Part 1

In [2]:
transactions = [
    {'A', 'B', 'D', 'G'},
    {'B', 'D', 'E'},
    {'A', 'B', 'C', 'E', 'F'},
    {'B', 'D', 'E', 'G'},
    {'A', 'B', 'C', 'E', 'F'},
    {'B', 'E', 'G'},
    {'A', 'C', 'D', 'E'},
    {'B', 'E'},
    {'A', 'B', 'E', 'F'},
    {'A', 'C', 'D', 'E'},
]
solver = APriori(transactions)

### All frequent items and rules
First we use the `get_frequent_item_sets` and `get_extracted_rules` methods to find all of the rules in the transactions that have `min_support` of 0.4 and `min_confidence` of 0.7.

In [3]:
all_frequents = solver.get_frequent_item_sets(.4)
all_rules = solver.get_extracted_rules(.7, frequent_item_set=all_frequents)
print('All frequent itemsets: (min_support = 0.4)')
for k in all_frequents:
    print('\n'.join(map(str, all_frequents[k])))
print()
print('All association rules: (min_confidence = 0.7)')
print('\n'.join(map(str, all_rules)))

All frequent itemsets: (min_support = 0.4)
{'A'}, s = 0.6
{'B'}, s = 0.8
{'C'}, s = 0.4
{'D'}, s = 0.5
{'E'}, s = 0.9
{'A', 'B'}, s = 0.4
{'C', 'A'}, s = 0.4
{'A', 'E'}, s = 0.5
{'B', 'E'}, s = 0.7
{'C', 'E'}, s = 0.4
{'D', 'E'}, s = 0.4
{'C', 'A', 'E'}, s = 0.4

All association rules: (min_confidence = 0.7)
{'C'} -> {'A'}, s = 0.4, c = 1.0
{'A'} -> {'E'}, s = 0.5, c = 0.8333333333333334
{'E'} -> {'B'}, s = 0.7, c = 0.7777777777777778
{'B'} -> {'E'}, s = 0.7, c = 0.875
{'C'} -> {'E'}, s = 0.4, c = 1.0
{'D'} -> {'E'}, s = 0.4, c = 0.8
{'A', 'E'} -> {'C'}, s = 0.4, c = 0.8
{'C', 'E'} -> {'A'}, s = 0.4, c = 1.0
{'C', 'A'} -> {'E'}, s = 0.4, c = 1.0
{'C'} -> {'A', 'E'}, s = 0.4, c = 1.0
{'C'} -> {'A', 'E'}, s = 0.4, c = 1.0


### a)
Frequent 1-itemsets:

In [4]:
print('\n'.join(map(str, all_frequents[1])))

{'A'}, s = 0.6
{'B'}, s = 0.8
{'C'}, s = 0.4
{'D'}, s = 0.5
{'E'}, s = 0.9


### b)
Frequent 2-itemsets:

In [5]:
print('\n'.join(map(str, all_frequents[2])))

{'A', 'B'}, s = 0.4
{'C', 'A'}, s = 0.4
{'A', 'E'}, s = 0.5
{'B', 'E'}, s = 0.7
{'C', 'E'}, s = 0.4
{'D', 'E'}, s = 0.4


### c)
Frequent 2-itemsets which generate a rule whose support value is greater than 0.7:

In [6]:
selected_rules = solver.get_extracted_rules(min_confidence=.7, frequent_item_set={2: all_frequents[2]})
print('\n'.join(map(lambda rule: str(rule.item_set), selected_rules)))

{'C', 'A'}
{'A', 'E'}
{'B', 'E'}
{'B', 'E'}
{'C', 'E'}
{'D', 'E'}


### d) 
Associan rules that have a confidence value of 1:

In [7]:
max_confident_rules = filter(lambda rule: rule.confidence == 1, all_rules)
print('\n'.join(map(str, max_confident_rules)))

{'C'} -> {'A'}, s = 0.4, c = 1.0
{'C'} -> {'E'}, s = 0.4, c = 1.0
{'C', 'E'} -> {'A'}, s = 0.4, c = 1.0
{'C', 'A'} -> {'E'}, s = 0.4, c = 1.0
{'C'} -> {'A', 'E'}, s = 0.4, c = 1.0
{'C'} -> {'A', 'E'}, s = 0.4, c = 1.0


### e)
As noted above, the `B => E` rule has a support of 0.7 and a confidence of 0.875.

## Part 2

In [9]:
transactions = [
    {'A', 'C', 'D'},
    {'B', 'C', 'E'},
    {'A', 'B', 'C', 'E'},
    {'B', 'E'}
]
solver = APriori(transactions)

### a)
Frequent itemsets with `min_support` of 0.4.

In [10]:
all_frequents = solver.get_frequent_item_sets(.4)
print('All frequent itemsets: (min_support = 0.4)')
for k in all_frequents:
    print('\n'.join(map(str, all_frequents[k])))

All frequent itemsets: (min_support = 0.4)
{'A'}, s = 0.5
{'B'}, s = 0.75
{'C'}, s = 0.75
{'E'}, s = 0.75
{'C', 'A'}, s = 0.5
{'C', 'B'}, s = 0.5
{'B', 'E'}, s = 0.75
{'C', 'E'}, s = 0.5
{'B', 'C', 'E'}, s = 0.5


## b)
Association rules from the previous itemsets with `min_confidence` bigger than 0.65.

In [14]:
selected_rules = solver.get_extracted_rules(.65, frequent_item_set=all_frequents)
print('Selected association rules: (min_confidence = 0.65)')
print('\n'.join(map(str, selected_rules)))

Selected association rules: (min_confidence = 0.65)
{'A'} -> {'C'}, s = 0.5, c = 1.0
{'C'} -> {'A'}, s = 0.5, c = 0.6666666666666666
{'B'} -> {'C'}, s = 0.5, c = 0.6666666666666666
{'C'} -> {'B'}, s = 0.5, c = 0.6666666666666666
{'E'} -> {'B'}, s = 0.75, c = 1.0
{'B'} -> {'E'}, s = 0.75, c = 1.0
{'E'} -> {'C'}, s = 0.5, c = 0.6666666666666666
{'C'} -> {'E'}, s = 0.5, c = 0.6666666666666666
{'C', 'E'} -> {'B'}, s = 0.5, c = 1.0
{'B', 'E'} -> {'C'}, s = 0.5, c = 0.6666666666666666
{'C', 'B'} -> {'E'}, s = 0.5, c = 1.0
{'E'} -> {'C', 'B'}, s = 0.5, c = 0.6666666666666666
{'C'} -> {'B', 'E'}, s = 0.5, c = 0.6666666666666666
{'E'} -> {'C', 'B'}, s = 0.5, c = 0.6666666666666666
{'B'} -> {'C', 'E'}, s = 0.5, c = 0.6666666666666666
{'B'} -> {'C', 'E'}, s = 0.5, c = 0.6666666666666666
{'C'} -> {'B', 'E'}, s = 0.5, c = 0.6666666666666666


### c)
Association rules from the previous itemsets with `min_confidence` bigger than 0.8.

In [17]:
selected_rules = solver.get_extracted_rules(.85, frequent_item_set=all_frequents)
print('Selected association rules: (min_confidence = 0.8)')
print('\n'.join(map(str, selected_rules)))

Selected association rules: (min_confidence = 0.8)
{'A'} -> {'C'}, s = 0.5, c = 1.0
{'E'} -> {'B'}, s = 0.75, c = 1.0
{'B'} -> {'E'}, s = 0.75, c = 1.0
{'C', 'E'} -> {'B'}, s = 0.5, c = 1.0
{'C', 'B'} -> {'E'}, s = 0.5, c = 1.0


### d)
As shown above, the confidence of `E => C` is 0.66.
### e)
As shown above, the confidence of `B => C` is 0.66.